src/ucs/memory/memtype_cache.c

```diff
@@ -201,23 +226,49 @@ UCS_PROFILE_FUNC(ucs_status_t, ucs_memtype_cache_lookup,
                  ucs_memtype_cache_t *memtype_cache, const void *address,
                  size_t size, ucs_memory_type_t *mem_type_p)
 {
-    const ucs_pgt_addr_t start = (uintptr_t)address;
+    ucs_pgt_addr_t start     = (uintptr_t)address;
+    const ucs_pgt_addr_t end = start + size;
+    ucs_status_t status      = UCS_OK;
+    int unkwn_mem_type       = 0;
     ucs_memtype_cache_region_t *region;
     ucs_pgt_region_t *pgt_region;
-    ucs_status_t status;
+
+    *mem_type_p = UCS_MEMORY_TYPE_LAST;
 
     pthread_rwlock_rdlock(&memtype_cache->lock);
 
-    pgt_region = UCS_PROFILE_CALL(ucs_pgtable_lookup, &memtype_cache->pgtable,
-                                  start);
-    if ((pgt_region == NULL) || (pgt_region->end < (start + size))) {
-        status = UCS_ERR_NO_ELEM;
-        goto out_unlock;
-    }
+    do {
+        pgt_region = UCS_PROFILE_CALL(ucs_pgtable_lookup,
+                                      &memtype_cache->pgtable,
+                                      start);
+        if (pgt_region == NULL) {
+            status = UCS_ERR_NO_ELEM;
+            goto out_unlock;
+        }
 
-    region      = ucs_derived_of(pgt_region, ucs_memtype_cache_region_t);
-    *mem_type_p = region->mem_type;
-    status      = UCS_OK;
+        region = ucs_derived_of(pgt_region, ucs_memtype_cache_region_t);
+        if (region->mem_type == UCS_MEMORY_TYPE_LAST) {
+            unkwn_mem_type = 1;
+        } else if (ucs_unlikely((*mem_type_p != UCS_MEMORY_TYPE_LAST) &&
+                                (*mem_type_p != region->mem_type))) {
+            ucs_error("0x%lx..0x%lx (mem_type %s) and 0x%lx..0x%lx "
```

Comment:
why needed to change the lookup function?

Reply:
we don't merge regions that have exact and unknown memory type,
but we want to report them to a user if a user's region consists of multiple regions (when at least one of them is unknown). we will report LAST and UCP will do memory type detection and then will update MemType Cache

Reply:
i don't think this code really works for host memory case
pls check

Reply:
@yosefe this is common code for all memory types
btw, we don't do a lookup for HOST buffers in gtests (since malloc() interceptors don't add them to MemType Cache), so can't test them. can only simulate it by generating HOST memory allocations

Reply:
for host memory only start address is checked, if not found - we return NO_ELEM immediately. without checking other regions which may intersect.
also what happens if we hit LAST memory type with start address and then CUDA memory? seems an error will not be reported, but if the order is reverse then it will be reported

Reply:
> also what happens if we hit LAST memory type with start address and then CUDA memory? seems an error will not be reported, but if the order is reverse then it will be reported

lookup returns  `UCS_OK` and `*mem_type_p=LAST` in both cases

Reply:
removed lookup changes as per offline discussion